File: solutions/9.palindrome-number.py
#
# @lc app=leetcode id=9 lang=python3
#
# [9] Palindrome Number
#

# @lc code=start
import math

class Solution:
    def isPalindrome(self, x: int) -> bool:
        # first
        # ------------------------
        # if x < 0:
        #     return False
        # stringX = str(x)
        # length = len(stringX)
        # halfLength = math.floor(length / 2)

        # if length % 2 == 0:
        #     return stringX[0:halfLength] == stringX[halfLength::][::-1]
        # else:
        #     return stringX[0:halfLength] == stringX[halfLength + 1::][::-1]

        # popular solution
        # ------------------------
        # if x < 0:
        #     return False

        # return str(x) == str(x)[::-1]

        # best solution
        # ------------------------
        if x < 0 or (x > 0 and x%10 == 0):
            return False
        result = 0
        while x > result:
            result = result * 10 + x % 10
            x = x // 10
        return True if (x == result or x == result // 10) else False

# @lc code=end

File: solutions/9.palindrome-number.ts
/*
 * @lc app=leetcode id=9 lang=typescript
 *
 * [9] Palindrome Number
 */

// @lc code=start
function isPalindrome(x: number): boolean {
  // population solution
  // -------------------
  // if(x < 0) {
  //   return false
  // }

  // return x === Number(String(x).split("").reverse().join(""))

  // best solution
  // -------------------
  if(x < 0 || (x > 0 && x % 10 === 0)) {
    return false;
  }

  let result = 0
  while(x > result) {
    result = result * 10 + x % 10
    x = Math.floor(x / 10)
  }

  return result === x || Math.floor(result / 10) === x
};
// @lc code=end
